cpp/src/arrow/dataset/discovery.cc

```diff
@@ -28,43 +28,82 @@
 #include "arrow/dataset/partition.h"
 #include "arrow/dataset/type_fwd.h"
 #include "arrow/filesystem/path_tree.h"
+#include "arrow/filesystem/path_util.h"
 #include "arrow/status.h"
 
 namespace arrow {
 namespace dataset {
 
 FileSystemDataSourceDiscovery::FileSystemDataSourceDiscovery(
-    fs::FileSystem* filesystem, std::string base_dir, std::vector<fs::FileStats> files,
-    std::shared_ptr<FileFormat> format)
+    fs::FileSystem* filesystem, std::vector<fs::FileStats> files,
+    std::shared_ptr<FileFormat> format, FileSystemDiscoveryOptions options)
     : fs_(filesystem),
-      base_dir_(std::move(base_dir)),
       files_(std::move(files)),
-      format_(std::move(format)) {}
+      format_(std::move(format)),
+      options_(std::move(options)) {}
 
-Status FileSystemDataSourceDiscovery::Make(fs::FileSystem* filesystem,
-                                           std::string base_dir,
-                                           std::vector<fs::FileStats> files,
-                                           std::shared_ptr<FileFormat> format,
-                                           std::shared_ptr<DataSourceDiscovery>* out) {
-  out->reset(
-      new FileSystemDataSourceDiscovery(filesystem, std::move(base_dir), files, format));
-  return Status::OK();
+bool StartsWithAnyOf(const std::vector<std::string>& prefixes, const std::string& path) {
+  auto dir_base = fs::internal::GetAbstractPathParent(path);
+  util::string_view basename{dir_base.second};
+
+  auto matches_prefix = [&basename](const std::string& prefix) -> bool {
+    return !prefix.empty() && basename.starts_with(prefix);
+  };
+
+  return std::none_of(prefixes.cbegin(), prefixes.cend(), matches_prefix);
```

Comment:
This should use `any_of` to match the function name, and the usage below should be changed to
```c++
      if (has_prefixes && StartsWithAnyOf(options.ignore_prefixes, path)) {
        continue;
      } 
```